### ISSUE TITLE
Show example of using raw html for browsers

### ISSUE THREAD

**Issue by Kali95739** [author/shell#12]
Can u show me an example of using a shell command in raw html for browser?

**Comment by Kali95739** [comment]
Something like

**Comment by Kali95739** [comment]
<script type=text/javascript import { Shell, Command } from '@author.io/shell'

// Define a command
const ListCommand = new Command({
  name: 'list',
  description: 'List the contents of the directory.',
  disableHelp: false, // Set to true to turn off default help messages for the entire shell (you can still provide your own). Defaults to false.
  // arguments are listed after the command in the default help screen. Ex: "dir list path"
  arguments: 'path', // Can be space/comma/tab/semicolon delimited or an array.
  alias: 'ls',
  // Any flag parsing options from the @author.io/arg library can be configured here.
  // See https://github.com/author/arg#configuration-methods for a list.
  flags: {
    long: {
      alias: 'l',
      description: 'Long format'.
      type: 'boolean',
      default: false
    },
    rootDir: {
      description: 'The root directory to list.',
      aliases: ['input', 'in', 'src'],
      single: true,
      // validate: RegExp/Function (see github.com/author/arg)
    }
  },
  handler (metadata, callback) {
    // ... this is where your command actually does something ...

    // Data comes from @author.io/arg lib. It looks like:
    // {
    //   command: <Command>,
    //   input: 'whatever user typed after "command"',
    //   flags: {
    //     recognized: {},
    //     unrecognized: [
    //       'whatever',
    //       'user',
    //       'typed'
    //     ]
    //   },
    //   valid: false,
    //   violations: [],
    //   flag (name) { return String },
    //   data (getter)
    // }
    console.log(metadata)

    // A single flag's value can be retrieved with this helper method.
    console.log(metadata.flag('long'))

    // Any unrecognized flags can be retrieved by index number (0-based)
    console.log(metadata.flag(0)) // The first unrecognized flag... returns null if it doesn't exist

    // Execution callbacks are optional. If a callback is passed from the
    // execution context to this handler, it will run after the command
    // has finished processing
    // (kind of like "next" in Express).
    // Promises are also supported.
    callback && callback()
  }
})

const shell = new Shell({
  name: 'myapp',
  version: '1.0.0',
  description: 'My demo app.',
  // This middleware runs before all command handlers.
  use: [
    (meta, next) => { ...; next() }
  ],
  // Trailers are like "post-middleware" that run after command handlers.
  trailer: [
    (meta, next) => { ...; next() }
    (meta, next) => { console.log('All done!') }
  ],
  commands: [
    // These can be instances of Command...
    list,

    // or just the configuration of a Command
    {
      name: 'find',
      description: 'Search metadoc for all the things.',
      alias: 'search',
      flags: {
        x: {
          type: 'string',
          required: true
        }
      },
      handler: (data, cb) => {
        console.log(data)
        console.log(`Mirroring input: ${data.input}`)

        cb && cb()
      }
      // Subcommands are supported
      // , commands: [...]
    }
  ]
})

// Run a command
shell.exec('find "some query"')

// Run a command using a promise.
shell.exec('find "some query"').then(() => console.log('Done!))

// Run a command using a callback (the callback is passed to the command's handler function)
shell.exec('find "some query"', () => console.log('Handled!'))

// Run a command, pass a callback to the handler, and use a promise to determine when everything is done.
shell.exec('find "some query"', () => console.log('Handled!')).then(() => console.log('Done!))

// Output the shell's default messages
console.log(shell.help)
console.log(shell.usage)
console.log(shell.description)</script>

**Comment by Kali95739** [comment]
Or something like that?

**Comment by coreybutler** [comment]
I'm sorry, I never saw this.

If I understand you correctly, you want to run a command from an HTML form. Here's an example:

https://gist.github.com/coreybutler/544b73e4d608bab97f0871136274daf9

**Comment by Kali95739** [comment]
Thank u but can we use this to run a powershell command for example?

**Comment by coreybutler** [comment]
@Kali95739 - this library is designed to create your own shell _programs_. You cannot directly execute a Powershell command from a webpage (security vulnerability, blocked by browsers). 

The only option is to have a web server running on the computer you want to run a command on. If this is your goal, you'll need a remote TTY library. Something like [tty.js](https://github.com/chjj/tty.js) or term.js are a better fit for that scenario.

**Comment by Kali95739** [comment]
I believe u can tbh but Its like a browser shim there is a way to module load I am sorry I thought this may have worked my mistake

**Comment by Kali95739** [comment]
But thank u

**Comment by coreybutler** [comment]
There is no way to run commands directly from a browser on a desktop. This has been explicitly prevented by browsers since the 90's. The "shim" you're referring to is a web server. Most people setup a web server and use socket.io (or some other web socket lib). From a web page, you can issue HTTP requests (or websocket "events") that the server receives/relays to the OS.

**Comment by Kali95739** [comment]
There is a way but I wont really try to change ur mind on it for one u
could use electron and there is more so

**Comment by Kali95739** [comment]
Pretty sure there is alot of things to transpile commonjs to es6 and then
use es6 in the browser

**Comment by coreybutler** [comment]
This has nothing to do with CJS, ES6, or transpilation. The browser explicitly prevents access to the underlying command system. If they did, then any batch/shell file could be run from the web. Viruses would run rampant, because that would be a _massive_ security breach. Even the Power APIs  (like access mic/camera/files) have limited access with forced approvals (prompts).

Electron has both Chromium and Node.js... two JS runtimes. The Chromium portion of Electron cannot communicate with the browser. Electron created a native IPC bridge (C++) linking the Chromium portion to Node.js. Node.js _does_ have access to the operating system, so it can execute commands. When a user inputs something in Electron, it would be relayed from Chromimum --> Node --> Desktop. This workaround is entirely because _browsers_ cannot do this.

Browser extensions have some more powerful features, but they still cannot interact with other programs on the desktop (they can read/write files).

There are ways to use a browser to issue a command, but browsers cannot do it on their own. Something else has to be installed on the computer to relay commands to the OS. Electron apps, Node, extensions, or other apps... all of those need to be installed separately. People do that all the time (see https://github.com/coreybutler/fenix or https://chrome.google.com/webstore/detail/cli/okpglddgmnblhbdpdcmodmacgcibgfkf), but it's not like a user can just go to a webpage and have a command run on their desktop.

**Comment by Kali95739** [comment]
What about something like nodular.js ?

**Comment by Kali95739** [comment]
Pretty sure it uses raw js to fetch the nodejs environment Im pretty sure u
can use it for child process

**Comment by coreybutler** [comment]
Nope. Nodular explicitly states it must be served by a web server (the exact same concept I outlined above).

<img width="614" alt="image" src="https://user-images.githubusercontent.com/770982/175796994-ac0abdab-a845-4fc6-82e3-ca8bc9a76034.png">

Running Node (or anything else) in the browser (via WASM) is explicitly sandboxed. Browsers will not execute untrusted code or commands on a computer, no exceptions (this is a legal liability for the browser makers). 

Browsers _can_ communicate with other apps via HTTP though. If people install a separate app with an HTTP server (API) to receive commands, the web page can communicate with it. That's why it is up to the other app to actually run the commands it receives.

**Comment by Kali95739** [comment]
Alright well thanks :)

**Comment by Kali95739** [comment]
There is even a es6 shim of child process

**Comment by coreybutler** [comment]
A shim just allows child_process to be imported in Node.js, i.e. `import child_process from 'child_process'`. It doesn't add runtime capabilities to other JavaScript runtimes (like the browser). Browsers do not have the child_process library. That's only part of the Node.js runtime.

Please read about [the Chrome security sandbox](https://chromium.googlesource.com/chromium/src/+/HEAD/docs/design/sandbox_faq.md#:~:text=The%20sandbox%20is,are%20sandboxed%20processes.), which clearly states the restrictions. You cannot run commands on a computer _directly_ from a browser.

Again, the workaround is to have another web-accessible application installed on the computer that does not have these restrictions, serving as a bridge/conduit for your browser code.
